integration-tests/application-server-integration-tests/src/test/java/co/elastic/apm/servlet/AbstractServletContainerIntegrationTest.java

```diff
@@ -340,39 +342,60 @@ private String getBaseUrl() {
 
     private List<JsonNode> getEvents(String eventType) {
         try {
-            final List<JsonNode> transactions = new ArrayList<>();
+            final List<JsonNode> events = new ArrayList<>();
             final ObjectMapper objectMapper = new ObjectMapper();
             for (HttpRequest httpRequest : mockServerContainer.getClient().retrieveRecordedRequests(request(INTAKE_V2_URL))) {
-                for (String ndJsonLine : httpRequest.getBodyAsString().split("\n")) {
+                final String bodyAsString = httpRequest.getBodyAsString();
+                validateEventMetadata(bodyAsString);
+                for (String ndJsonLine : bodyAsString.split("\n")) {
                     final JsonNode ndJson = objectMapper.readTree(ndJsonLine);
                     if (ndJson.get(eventType) != null) {
-                        transactions.add(ndJson.get(eventType));
+                        events.add(ndJson.get(eventType));
                     }
                 }
             }
-            return transactions;
+            return events;
         } catch (IOException e) {
             throw new RuntimeException(e);
         }
     }
 
-    public void validateMetadata() {
+    private void validateEventMetadata(String bodyAsString) {
         try {
             final ObjectMapper objectMapper = new ObjectMapper();
-            final JsonNode payload;
-            payload = objectMapper
-                .readTree(mockServerContainer.getClient()
-                    .retrieveRecordedRequests(request(INTAKE_V2_URL))[0].getBodyAsString().split("\n")[0]);
-            JsonNode metadata = payload.get("metadata");
-            assertThat(metadata.get("service").get("name").textValue()).isEqualTo(expectedDefaultServiceName);
-            JsonNode container = metadata.get("system").get("container");
-            assertThat(container).isNotNull();
-            assertThat(container.get("id").textValue()).isEqualTo(servletContainer.getContainerId());
+            for (String line : bodyAsString.split("\n")) {
+                final JsonNode event = objectMapper.readTree(line);
+                final JsonNode metadata = event.get("metadata");
+                if (metadata != null) {
+                    validataMetadataEvent(metadata);
+                } else {
+                    // TODO make it work ;)
+                    //  validateServiceName(event.get("error"));
+                    //  validateServiceName(event.get("span"));
```

Comment:
Working now? 🙂

Reply:
Nope, this is still a WIP. The problem is when there's a span being created in another thread. In that case, the span does not have a reference to its transaction and thus can't copy the service_name. I would like to avoid looking up the service name for each span again for these reasons:

* Performance
  It is cheaper if we only need to lookup once per transaction
* API surface
  Requiring to provide the initiating class loader for each span creation would greatly increase the number of changes required to make this happen
* Increased chance of wrong service_names
  When creating a JDBC span, for example, we don't really know the actual initiating class loader. That is because we don't need the CL of the JDBC implementation but the web application class loader. Especially when the application server ships with the JDBC implementation, the class loaders would be different.

Therefore, I think we need to have the ability to add tags to the `TraceContext`. Those would be inherited to child contexts. This ensures a consistent service_name when creating child spans of already ended transactions. Maybe we don't even need support for generic tags for now but just move the `serviceName` to `TraceContext`.

Reply:
Ahh, good point.
I assume the latter is a better option, just like `TraceContext` inherits other things like IDs and flags. Not sure generifying it is necessary

Reply:
done